### PR TITLE
Added CLAHE adjustment the all the anatomical canvases.

### LINKED ISSUE #1660: brightness/contrast features for sct_label_utils -create-viewer
There are some very dark images in which it is very difficult/imposible to label.
Proposal:
- Include a slice control bar to adjust the brightness
- Include a slice bar control to adjust the contrast
- Include a check box, when active performs an automatic histogram adjustment, for example adaptive equalization, see exposure.equalize_adapthist () function in scikit-image, link below
http://scikit-image.org/docs/dev/auto_examples/color_exposure/plot_equalize.html

### PR DESCRIPTION
**Pull request by peristeri**
Fixes #1660

### CODE REVIEW

File: spinalcordtoolbox/gui/widgets.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from matplotlib.backends.backend_qt4agg import NavigationToolbar2QT as NavigationToolbar
 from matplotlib.figure import Figure
 from matplotlib.widgets import Cursor
+from skimage import exposure
```
Comment (zougloub) [inline spinalcordtoolbox/gui/widgets.py]: Importing `skimage.exposure` and using the fully qualified function name would be more explicit.
But that's okay.

File: spinalcordtoolbox/gui/widgets.py [hunk 2]
```diff
@@ -139,13 +140,9 @@ def _init_ui(self, data, aspect):
...
         self._axes.axis('off')
         self.view = self._axes.imshow(
-            data,
+            self.adjust_image(data),
             cmap=self._params.cmap,
-            interpolation=self._params.interp,
-            vmin=self._params.vmin,
-            vmax=self._params.vmax,
-            alpha=self._params.alpha)
-        self._axes.set_aspect(aspect)
```
Comment (jcohenadad) [inline spinalcordtoolbox/gui/widgets.py]: Why removing it again?  It is needed for anisotropic resolution. I fixed this bug in a [recent commit](https://github.com/neuropoly/spinalcordtoolbox/pull/1545/commits/d9704a46d14fc53e16150708a3c57cd2c3f90f8e), and now you are reintroducing it without justification. 

Without self._axes.set_aspect(aspect):
screen shot 2018-03-29 at 11 16 58

With self._axes.set_aspect(aspect):
screen shot 2018-03-29 at 11 18 25

File: spinalcordtoolbox/gui/widgets.py (hunk 1 above)
Comment (jcohenadad) [reply spinalcordtoolbox/gui/widgets.py]: fixed in d5c4b017289f742ff52562cebdc264727446543b

**Comment by jcohenadad** [comment]
bug:
~~~
~/sct_example_data/mt $ sct_label_utils -i mt0.nii.gz -create-viewer 2,4

--
Spinal Cord Toolbox (image-adjustment/8f288717a815db14d6d2c39a3f31e0a96bc5e262)
Running /Users/julien/code/sct/scripts/sct_label_utils.py -i mt0.nii.gz -create-viewer 2,4
/Users/julien/code/sct/python/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_label_utils.py", line 824, in <module>
    main()
  File "/Users/julien/code/sct/scripts/sct_label_utils.py", line 814, in main
    processor.process(process_type)
  File "/Users/julien/code/sct/scripts/sct_label_utils.py", line 112, in process
    self.output_image = self.launch_sagittal_viewer(self.value)
  File "/Users/julien/code/sct/scripts/sct_label_utils.py", line 653, in launch_sagittal_viewer
    launch_sagittal_dialog(self.image_input, output, params)
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/sagittal.py", line 100, in launch_sagittal_dialog
    dialog = SagittalDialog(controller)
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/base.py", line 110, in __init__
    self._init_ui()
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/base.py", line 118, in _init_ui
    self._init_canvas(layout)
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/sagittal.py", line 52, in _init_canvas
    self.sagittal = widgets.SagittalCanvas(self, plot_points=True, annotate=True)
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/widgets.py", line 243, in __init__
    self._init_ui(self._image.data[:, :, self._z], dx / dy)
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/widgets.py", line 143, in _init_ui
    self.adjust_image(data),
  File "/Users/julien/code/sct/spinalcordtoolbox/gui/widgets.py", line 177, in adjust_image
    return exposure.equalize_adapthist(data)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/skimage/color/adapt_rgb.py", line 40, in image_filter_adapted
    return image_filter(image, *args, **kwargs)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/skimage/exposure/_adapthist.py", line 76, in equalize_adapthist
    image = img_as_uint(image)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/skimage/util/dtype.py", line 325, in img_as_uint
    return convert(image, np.uint16, force_copy)
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/skimage/util/dtype.py", line 205, in convert
    raise ValueError("Images of type float must be between -1 and 1.")
ValueError: Images of type float must be between -1 and 1.
~~~

**Comment by jcohenadad** [comment]
Another issue, already reported [here](https://github.com/scikit-image/scikit-image/issues/2219), is that CLAHE introduces an artifact when the kernel size is not a multiple of the image dimensions. See below:

WITHOUT CLAHE:
screen shot 2018-03-29 at 14 38 31

WITH CLAHE:
screen shot 2018-03-29 at 14 38 09

**Comment by jcohenadad** [comment]
important bug, please do NOT merge before fixing it.

data: sct_example_data/t2
syntax:
~~~
sct_label_vertebrae -i t2.nii.gz -s t2_seg.nii.gz -c t2 -initc2
~~~

master/caf11f46d92fe8765490a53325acbf0e2b36b8f1
screen shot 2018-04-24 at 12 59 17

image-adjustment/6fb676cbde7bf70ee7d3a5490cbc6e705774665c:
screen shot 2018-04-24 at 13 00 06

**Comment by jcohenadad** [comment]
closing this because:
- no activity for a month
- manual adjustment of contrast is now possible, which makes this PR less required

**Pull request closed without merging**